### PR TITLE
Fix stream helpers

### PR DESCRIPTION
**Pull request by bergie**
The stream helpers from #467 don't seem to behave correctly. Need to be reimplemented.

Also dropping the `input.buffer` interface since it encourages bypassing normal input handling which is necessary for bracket forwarding and activation/deactivation.

### CODE REVIEW

File: src/lib/Component.coffee [hunk 1]
```diff
@@ -358,17 +357,30 @@ class ProcessInput
...
+        if ip.type is 'data'
+          # Data IP on its own is a valid stream
+          return true unless portBrackets.length
+          # Otherwise we need to check for complete stream
+          hasData = true
+          return false
+        if ip.type is 'closeBracket'
+          portBrackets.pop()
+          return false if portBrackets.length
+          return false unless hasData
```
Comment (trustmaster) [inline src/lib/Component.coffee]: Thus it disallows empty streams like `< < > >`. What's the reason behind it? My guess is bracket forwarding not working correctly for empty streams. Should be documented somewhere

File: src/lib/Component.coffee [hunk 2]
```diff
@@ -358,17 +357,30 @@ class ProcessInput
...
+          return true
+      return false unless @ports[port].has @scope, validate
+    res = true
+    res and= @ports[port].ready @scope for port in args
```
Comment (trustmaster) [inline src/lib/Component.coffee]: What's this line for if it's already checked with `InPort.has()`?

File: src/lib/Component.coffee [hunk 3]
```diff
@@ -461,64 +473,39 @@ class ProcessInput
...
+        if ip.type is 'data'
+          portPackets.push ip
+          hasData = true
+          # Unbracketed data packet is a valid stream
+          break unless portBrackets.length
+        if ip.type is 'closeBracket'
+          portPackets.push ip
+          portBrackets.pop()
+          if hasData and not portBrackets.length
+            # Last open bracket finishes stream
```
Comment (trustmaster) [inline src/lib/Component.coffee]: Typo here, should be `Last close bracket`

File: src/lib/Component.coffee (hunk 1 above)
Comment (bergie) [inline src/lib/Component.coffee]: Yes, this is based on what we agreed as being a "stream" in the meetup last summer. Either:

* Data packet not surrounded by brackets
* Data packet(s) surrounded by matching brackets

So empty brackets are not considered a stream.

**Pull request merged**